src/mbgl/style/conversion/function.cpp

```diff
@@ -0,0 +1,702 @@
+    case FunctionType::Exponential:
+        return convertExponentialFunction(type, value, error, number(get(literal(*property))));
+    case FunctionType::Categorical:
+        return convertCategoricalFunction(type, value, error, *property);
+    case FunctionType::Identity:
+        return type.match(
+            [&] (const type::StringType&) -> optional<std::unique_ptr<Expression>> {
+                return string(get(literal(*property)));
+            },
+            [&] (const type::NumberType&) -> optional<std::unique_ptr<Expression>> {
+                return number(get(literal(*property)));
+            },
+            [&] (const type::BooleanType&) -> optional<std::unique_ptr<Expression>> {
+                return boolean(get(literal(*property)));
+            },
+            [&] (const type::ColorType&) -> optional<std::unique_ptr<Expression>> {
+                return toColor(get(literal(*property)));
+            },
+            [&] (const type::Array& array) -> optional<std::unique_ptr<Expression>> {
+                return std::unique_ptr<Expression>(
+                    std::make_unique<ArrayAssertion>(array, get(literal(*property))));
+            },
+            [&] (const auto&) -> optional<std::unique_ptr<Expression>>  {
+                assert(false); // No properties use this type.
+                return {};
+            }
+        );
+    default:
+        error = { "unsupported function type" };
+        return {};
+    }
+}
+
+template <class T>
+optional<std::unique_ptr<Expression>> composite(type::Type type,
+                                                const Convertible& value,
+                                                Error& error,
+                                                std::unique_ptr<Expression> (*makeInnerExpression) (type::Type type,
+                                                                                                    double base,
+                                                                                                    const std::string& property,
+                                                                                                    std::map<T, std::unique_ptr<Expression>>)) {
+    auto propertyValue = objectMember(value, "property");
+    if (!propertyValue) {
+        error = { "function must specify property" };
+        return {};
+    }
+
+    auto base = convertBase(value, error);
+    if (!base) {
+        return {};
+    }
+
+    auto propertyString = toString(*propertyValue);
+    if (!propertyString) {
+        error = { "function property must be a string" };
+        return {};
+    }
+
+    auto stopsValue = objectMember(value, "stops");
+    if (!stopsValue) {
+        error = { "function value must specify stops" };
+        return {};
+    }
+
+    if (!isArray(*stopsValue)) {
+        error = { "function stops must be an array" };
+        return {};
+    }
```

Comment:
This check and the previous one are already happening in `convertCompositeFunctionToExpression`